hub/api/tensor.py

```diff
@@ -1,35 +1,68 @@
 from hub.util.slice import merge_slices
 import numpy as np
 from typing import Union
+from hub.core.storage.provider import StorageProvider
+from hub.core.storage.memory import MemoryProvider
+from hub.core.storage.local import LocalProvider
+from hub.core.chunk_engine.read import read_array, read_tensor_meta
+from hub.core.chunk_engine.write import write_array, write_tensor_meta
 
 
 class Tensor:
-    def __init__(self, uuid: str, tensor_slice: slice = slice(None)):
+    def __init__(
+        self,
+        key: str,
+        provider: StorageProvider,
+        tensor_slice: slice = slice(None),
+    ):
         """Initialize a new tensor.
 
         Note:
             This operation does not create a new tensor in the backend,
             and should normally only be performed by Hub internals.
 
         Args:
-            uuid (str): The internal identifier for this tensor.
-            tensor_slice (slice, optional): The slice object restricting the view of this tensor.
+            key (str): The internal identifier for this tensor.
+            provider (StorageProvider): The storage provider for the parent dataset.
+            tensor_slice (slice): The slice object restricting the view of this tensor.
         """
-        self.uuid = uuid
+        self.key = key
+        self.provider = provider
         self.slice = tensor_slice
-        self.shape = (0,)  # Dataset should pass down relevant metadata
+
+        self._update_from_meta()
+
+    def _update_from_meta(self):
+        meta = read_tensor_meta(self.key, self.provider)
+        self.num_samples = meta["length"]
+        self.shape = meta["max_shape"]
 
     def __len__(self):
         """Return the length of the primary axis"""
-        return self.shape[0]
+        return self.num_samples
 
     def __getitem__(self, item: Union[int, slice]):
         if isinstance(item, int):
             item = slice(item, item + 1)
 
         if isinstance(item, slice):
             new_slice = merge_slices(self.slice, item)
-            return Tensor(self.uuid, new_slice)
+            return Tensor(self.key, self.provider, new_slice)
+
+    def __setitem__(self, item: Union[int, slice], value: np.ndarray):
+        sliced_self = self[item]
+        if sliced_self.slice != slice(None):
```

Comment:
Why do we have this error?

Reply:
This is in `__setitem__`. It's a separate task to implement updating a subsection of a chunk, so I just throw for now. It would happen in the following instance:
```
ds = Dataset(...)
ds["image"] = np.zeros((10000, 28, 28))
ds["image"][300:400] = np.ones((100, 28, 28)) # Sliced update not supported. Throws.
```